### ISSUE TITLE
Conformity to Rust API guidelines

### ISSUE THREAD

**Issue by JSorngard** [DomiDre/gauss-quad#27]
This is an issue that tracks the process of making the crate's API conform to the Rust API guidelines: https://rust-lang.github.io/api-guidelines/.
A ⚠️ in front if a task means that implementing it is a breaking change and would require a major version bump.

 - [x] ⚠️ The constructor should probably be named `new` instead of `init`: https://rust-lang.github.io/api-guidelines/predictability.html#c-ctor and https://rust-lang.github.io/api-guidelines/interoperability.html#types-eagerly-implement-common-traits-c-common-traits. Completed in #35.
 - [x] The quadrature rule structs should probably implement the `Serialize` and `Deserialize` traits from serde. For this crate that should probably be behind a feature flag in order to not slow down compile times for those who don't need it: https://rust-lang.github.io/api-guidelines/interoperability.html#data-structures-implement-serdes-serialize-deserialize-c-serde. Completed in #38.
 - [x] All constructors that can panic should have a `# Panics` section in their docs: https://rust-lang.github.io/api-guidelines/documentation.html#function-docs-include-error-panic-and-safety-considerations-c-failure. Completed in #33.
 - [x] The `keywords` and `categories` fields should probably be added to `Cargo.toml`:  https://rust-lang.github.io/api-guidelines/documentation.html#cargotoml-includes-all-common-metadata-c-metadata. Completed in #28 #32.
 - [x] ⚠️ The `nodes` and `weights` fields on the quadrature rule structs are not private. Right now the user can edit them accidentally and then get the wrong integration result. Instead the crate could supply functions like `pub fn nodes(&self) -> &[f64] {&self.nodes}` that make the data accessible without allowing anyone to accidentally edit it and remove the invariants. If the crate is to be used from other languages it may be necessary to add `#[repr(C)]` to the structs, though I don't know: https://rust-lang.github.io/api-guidelines/future-proofing.html#structs-have-private-fields-c-struct-private. Completed in #46.

**Comment by DomiDre** [comment]
Oi good job collecting those. I love abiding to guide lines so these fixes should be made. Didn't know of that guide yet, so good to know. If we have to bump to v0.2.0, so be it. I think a CHANGELOG.md should then also be added to the crate for the breaking changes.
I could tackle the boxes in the next evenings (if you dont get to it beforehand). For all the work you've done already, I'll also add you to the repo. Can also add you to the authors list in the Cargo.toml if you want.

**Comment by JSorngard** [comment]
@DomiDre I'd love to be listed among the authors, and thank you for the invitation!

It sounds like a good idea to make a changelog if we do breaking changes. Maybe we start with the non-breaking ones first?

**Comment by DomiDre** [comment]
For the remaining issues I open a branch v0.2.0 to branch off from and merge into until the new API is stabilized, or?

**Comment by JSorngard** [comment]
Yeah, that sounds good.  
It seems reasonable to implement the serialization only after the API change since I don't know exactly how different versions of the crate trying to deserialize from each others data would go if the data is structured differently between them.

**Comment by JSorngard** [comment]
Nice! Now we have all of that done 🎉

Might it be a good idea to take a bit of a pause before we release in case we find something small that we want to change?

**Comment by DomiDre** [comment]
Sounds good. Maybe I will take a look if I can find a use case for parallel integration and if it's really worth integrating into the crate.

**Comment by JSorngard** [comment]
Everything in this issue is completed! 🎉
